src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -193,37 +186,98 @@ public async Task UpdatePrimaryBranchSolutionAsync(AssetProvider assetProvider,
-                {
                     return availableSolution;
-                }
 
-                var solution = await CreateSolution_NoLockAsync(
+                var solution = await CreateFullSolution_NoLockAsync(
                     assetProvider,
                     solutionChecksum,
                     fromPrimaryBranch,
                     workspaceVersion,
                     CurrentSolution,
                     cancellationToken).ConfigureAwait(false);
 
-                _lastRequestedSolutionWithChecksum = Tuple.Create(solutionChecksum, solution);
-
+                _lastRequestedSolutionWithChecksum = new(solutionChecksum, solution);
                 return solution;
             }
         }
 
+        private ValueTask<Solution> GetProjectSubsetSolutionAsync(
+            AssetProvider assetProvider,
+            Checksum solutionChecksum,
+            ProjectId projectId,
+            CancellationToken cancellationToken)
+        {
+            // Attempt to just read without incurring any other costs.
+            if (_lastRequestedProjectIdToSolutionWithChecksum.TryGetValue(projectId, out var tuple) &&
+                tuple.Item1 == solutionChecksum)
+            {
+                return new(tuple.Item2);
+            }
+
+            return GetProjectSubsetSolutionSlowAsync(tuple?.Item2 ?? CurrentSolution, assetProvider, solutionChecksum, projectId, cancellationToken);
+
+            async ValueTask<Solution> GetProjectSubsetSolutionSlowAsync(
+                Solution baseSolution,
+                AssetProvider assetProvider,
+                Checksum solutionChecksum,
+                ProjectId projectId,
+                CancellationToken cancellationToken)
+            {
+                var updater = new SolutionCreator(Services.HostServices, assetProvider, baseSolution, cancellationToken);
+
+                // check whether solution is update to the given base solution
+                Solution result;
+                if (await updater.IsIncrementalUpdateAsync(solutionChecksum).ConfigureAwait(false))
+                {
+                    // create updated solution off the baseSolution
+                    result = await updater.CreateSolutionAsync(solutionChecksum).ConfigureAwait(false);
+                }
+                else
+                {
+                    // we need new solution. bulk sync all asset for the solution first.
+                    await assetProvider.SynchronizeSolutionAssetsAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+
+                    // get new solution info and options
+                    var (solutionInfo, options) = await assetProvider.CreateSolutionInfoAndOptionsAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+
+                    var workspace = new TemporaryWorkspace(Services.HostServices, WorkspaceKind.RemoteTemporaryWorkspace, solutionInfo, options);
+                    result = workspace.CurrentSolution;
```

Comment:
projectId not used here?

Reply:
no.  it's not really needed (unless i'm missing something). w e coudl assert that the solution that comes back at least has that project in it.  but beyond that, i'm not sure howe else we would use it.

Reply:
In that case maybe return (solutionChecksum, result) and store it to _lastRequestedProjectIdToSolutionWithChecksum in the caller - then you can remove the projectId param.